<div style="text-align: justify; padding:5px; background-color:rgb(252, 253, 255); border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    <font color='red'>Mini Jupyter tutorial<br><br>To run each cell, click the cell and press <kbd>Run</kbd> from the menu bar. This will run any Python code or display any text within the selected cell before highlighting the next cell down. There are two types of cell: A <i>text cell</i> of type <kbd>Markdown</kbd> or <kbd>Heading</kbd> and a <i>code cell</i> of type <kbd>Code</kbd> identifiable with the <span style="font-family: courier; color:black; background-color:white;">In[ ]:</span> to the left of the cell</i>. The type of cell is also identifiable from the dropdown menu in the above menu bar to the right of <kbd>Run</kbd>. Any visual results produced by the code (text/figures) are displayed directly below that cell. Press <kbd>Run</kbd> again until you reach the end of the notebook or alternatively click <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart and Run All</kbd>. Should the Jupyter notebook crash for any reason, restart the Jupyter Kernel by clicking <kbd>Kernel</kbd><font color='black'>→</font><kbd>Restart</kbd>, and start again from the top.
        
</div>

# Tutorial 2.6: Metabolomics of Gastric Cancer: A multi-block hierarchical edge bundle

<p style="text-align: justify">
<br>
This tutorial covers the necessary steps for producing a Hierarchical Edge Bundle using multi-block data from a gastric cancer study.
</p>

<div style="text-align: justify; padding:5px; background-color:rgb(252, 253, 255); border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    <font color='red', size=4>Note: If visualisng using a JavaScript pop-up window you will need to allow pop-ups from your browser for the domain you're running from (localhost or mybinder.org).
</div> 

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">
    
<h2 id="1importpackagesmodules" style="text-align: justify">1. Import Packages/Modules</h2>

<p style="text-align: justify">The first code cell of this tutorial imports <a href="https://docs.python.org/3/tutorial/modules.html"><em>packages</em> and <em>modules</em></a> into the Jupyter environment. <em>Packages</em> and <em>modules</em> provide additional functions and tools beyond the in-built Python modules.
<br></p>
<br>
All the code embedded in this notebook is written using Python (<a href="http://www.python.org">python.org</a>) and JavaScript (<a href="https://www.javascript.com/">javascript.com</a>) and are built upon popular open source packages such as Networkx (<a href="https://networkx.github.io/">networkx.github.io</a>), NumPy (<a href="https://numpy.org/">numpy.org</a>), SciPy (<a href="https://www.scipy.org/">scipy.org</a>), Matplotlib (<a href="https://matplotlib.org/">matplotlib.org</a>), statsmodels (<a href="www.statsmodels.org/">statsmodels.org</a>), Scikit-learn (<a href="scikit-learn.org/">scikit-learn.org</a>), scikits.bootstrap (<a href="github.com/cgevans/scikits-bootstrap">github.com/cgevans/scikits-bootstrap</a>), Pandas (<a href="https://pandas.pydata.org/">pandas.pydata.org</a>) and D3 JavaScript (<a href="https://d3js.org/">d3js.org</a>).
    
<em>Note:</em> a tutorial focusing on the python programming language is beyond the scope of this notebook. To learn how to program in Python with Jupyter Notebook please refer to: 
<a href="https://mybinder.org/v2/gh/jakevdp/PythonDataScienceHandbook/master?filepath=notebooks%2FIndex.ipynb">Python Data Science Handbook (Jake VanderPlas, 2016)</a>.

In [1]:
import os
    
home = os.getcwd() + "/"

import numpy as np
import pandas as pd
from IPython.display import Javascript, display, IFrame
import multivis

print('All packages successfully loaded')

%load_ext autoreload
%autoreload 2

All packages successfully loaded


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

<h2 style="text-align: justify">2. Load Data and Peak Table</h2>

<p style="text-align: justify">The code cell below loads the <em>Data</em> and <em>Peak</em> tables from an Excel file using <code>loadData()</code>. When this is complete, you should see confirmation that Peak (the Peak worksheet) and Data (the Data worksheet) tables have been loaded.<br>

This dataset has previously been published in (<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4716538/">Chan, A, et al. (2016)</a>) in <i>British Journal of Cancer</i> and has been put into a standardised <a href="https://en.wikipedia.org/wiki/Tidy_data">Tidy Data</a> format.
</p>

Please inspect the <a href="GastricCancer_NMR.xlsx">GastricCancer_NMR.xlsx </a>Excel file before using it in this tutorial to understand its structure. To change the dataset to be loaded into the notebook replace <code>filename = 'GastricCancer_NMR.xlsx'</code> with another file with the same <a href="https://en.wikipedia.org/wiki/Tidy_data">Tidy Data</a> format as <a href="GastricCancer_NMR.xlsx">GastricCancer_NMR.xlsx</a>, and then rerun the workflow.

</div></div>

In [2]:
file = 'GastricCancer_NMR.xlsx'

DataTable,PeakTable = multivis.utils.loadData(home + file, DataSheet='Data', PeakSheet='Peak')

Loading table: Peak
Loading table: Data
TOTAL SAMPLES: 123 TOTAL PEAKS: 149
Done!


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Data Table

Check the imported Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(DataTable)</span><br>
</div>

In [3]:
display(DataTable)

Idx    SampleID SampleType                   Class     M1       M2  \
1      2    sample_2     Sample          Gastric Cancer   43.0    525.7   
2      3    sample_3     Sample  Benign Gastric Disease  214.3  10703.2   
3      4    sample_4     Sample                 Healthy   31.6     59.7   
4      5    sample_5     Sample          Gastric Cancer   81.9    258.7   
5      6    sample_6     Sample  Benign Gastric Disease  196.9    128.2   
..   ...         ...        ...                     ...    ...      ...   
119  134  sample_134     Sample  Benign Gastric Disease  133.9    172.6   
120  135  sample_135     Sample                 Healthy    7.5    390.5   
121  137  sample_137     Sample          Gastric Cancer  405.3    510.7   
122  138  sample_138     Sample  Benign Gastric Disease   45.4    191.6   
123  139  sample_139     Sample                 Healthy   30.7     56.8   

        M3     M4     M5     M6  ...   M140    M141   M142    M143   M144  \
1    130.2    NaN  694.5  114.5  ...   84.2   357.1   16.1   455.5   29.5   
2    104.7   46.8  483.4  152.3  ...  993.5  1698.5   32.9    75.9   33.2   
3     86.4   14.0   88.6   10.3  ...   58.1    83.5   60.5   136.9   17.0   
4    315.1    8.7  243.2   18.4  ...   44.5    47.6   45.6  1441.7   35.2   
5    862.5   18.7  200.1    4.7  ...  143.8   157.2   10.4   182.1   32.6   
..     ...    ...    ...    ...  ...    ...     ...    ...     ...    ...   
119  121.5  128.4  324.6   79.7  ...   69.3   268.4    7.1    62.7   27.4   
120   67.9   38.4   58.9   26.5  ...   99.7   184.4    1.7    94.7   32.3   
121  521.9   91.9  732.1  145.7  ...  434.8    84.8  182.3   110.7  123.9   
122   41.0   18.7   40.8   32.2  ...   45.3    44.5   14.5    83.8   27.9   
123   35.9   20.9   17.4    NaN  ...   28.0    38.7    1.3   130.9   24.6   

      M145   M146   M147    M148   M149  
1     28.1   35.8  316.1   390.7  199.0  
2    802.8  967.6  154.4    31.6  195.2  
3     10.2   24.7   64.1    91.4   91.6  
4      0.1   22.8  135.0   322.3  254.3  
5    435.1  325.3  162.4   129.7  207.2  
..     ...    ...    ...     ...    ...  
119    2.8   32.5  249.6  1868.7  325.8  
120    NaN    5.8  174.2     NaN  145.5  
121    0.4   36.3   60.1   317.3  401.7  
122    0.3    0.5   47.3    47.8   46.5  
123    0.7    5.9   20.7   124.1   28.9  

[123 rows x 153 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table

Check the imported Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTable)</span><br>
</div>

In [4]:
display(PeakTable)

Idx  Name                    Label     Score      VIP1    Pvalue  \
1      1    M1        1_3-Dimethylurate  0.273432  0.921751  0.899999   
2      2    M2  1_6-Anhydro-β-D-glucose  0.386187  2.071139  0.586879   
3      3    M3     1_7-Dimethylxanthine  0.122120  2.892664  0.068365   
4      4    M4     1-Methylnicotinamide  0.781881  2.360494  0.877783   
5      5    M5           2-Aminoadipate  0.412730  2.969282  0.312104   
..   ...   ...                      ...       ...       ...       ...   
145  145  M145                    uarm1  0.226951  0.805725  0.584023   
146  146  M146                    uarm2  0.370446  2.261353  0.882148   
147  147  M147                β-Alanine  0.853027  2.679534  0.679136   
148  148  M148        π-Methylhistidine  0.116367  2.554018  0.839549   
149  149  M149        τ-Methylhistidine  0.870186  0.739459  0.445225   

        QC_RSD  
1    32.208005  
2    31.178028  
3    34.990605  
4    12.804201  
5     9.372664  
..         ...  
145  41.406985  
146  34.458172  
147  27.623517  
148  16.561921  
149   8.351801  

[149 rows x 7 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 3. Calculate the percentage of missing values in the data

</div>

In [5]:
stats_missing = multivis.utils.statistics(PeakTable, DataTable)

stats_missing.help()

Generate a table of parametric or non-parametric statistics and merges them with the Peak Table (node table).
        Initial_Parameters
            ----------
            peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
            datatable : Pandas dataframe matrix containing values for statistical analysis

        Methods
            -------
            set_params : Set parameters -
                parametric: Perform parametric statistical analysis, assuming the data is normally distributed (default: True)
                log_data: Perform a log ('natural', base 2 or base 10) on all data prior to statistical analysis (default: (False, 2))
                scale_data: Scale the data to unit variance (default: False)
                impute_data: Impute any missing values using KNN impute with a set number of nearest neighbours (default: (False, 3))
                group_column_name: The group column name used in the datatable (default: None)
      

In [6]:
params = dict({'parametric': False
              , 'log_data': (False, 10)
              , 'group_column_name': 'Class'
              , 'control_group_name': 'Healthy'
              , 'group_alpha_CI': 0.05
              , 'fold_change_alpha_CI': 0.05
              , 'pca_alpha_CI': 0.05
              , 'total_missing': True
              , 'group_missing': True
              , 'pca_loadings': False
              , 'normality_test': False
              , 'group_normality_test': False
              , 'group_mean_CI': False
              , 'group_median_CI': False
              , 'mean_fold_change': False
              , 'median_fold_change': False
              , 'kruskal_wallis_test': False
              , 'levene_twoGroup': False
              , 'levene_allGroup': False
              , 'oneway_Anova_test': False
              , 'ttest_oneGroup': False
              , 'ttest_twoGroup': False
              , 'mann_whitney_u_test': False})

stats_missing.set_params(**params)

PeakTablePercentMissing = stats_missing.calculate()

In [7]:
display(PeakTablePercentMissing)

Idx  Name                    Label     Score      VIP1    Pvalue  \
0      0    M1        1_3-Dimethylurate  0.273432  0.921751  0.899999   
1      1    M2  1_6-Anhydro-β-D-glucose  0.386187  2.071139  0.586879   
2      2    M3     1_7-Dimethylxanthine  0.122120  2.892664  0.068365   
3      3    M4     1-Methylnicotinamide  0.781881  2.360494  0.877783   
4      4    M5           2-Aminoadipate  0.412730  2.969282  0.312104   
..   ...   ...                      ...       ...       ...       ...   
144  144  M145                    uarm1  0.226951  0.805725  0.584023   
145  145  M146                    uarm2  0.370446  2.261353  0.882148   
146  146  M147                β-Alanine  0.853027  2.679534  0.679136   
147  147  M148        π-Methylhistidine  0.116367  2.554018  0.839549   
148  148  M149        τ-Methylhistidine  0.870186  0.739459  0.445225   

        QC_RSD  Percent_Group_Benign Gastric Disease_Missing  \
0    32.208005                                          10.0   
1    31.178028                                           0.0   
2    34.990605                                           7.5   
3    12.804201                                           7.5   
4     9.372664                                           5.0   
..         ...                                           ...   
144  41.406985                                          10.0   
145  34.458172                                           2.5   
146  27.623517                                           0.0   
147  16.561921                                           0.0   
148   8.351801                                           0.0   

     Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  \
0                               11.627907                           12.5   
1                                2.325581                            0.0   
2                                2.325581                            7.5   
3                               16.279070                            5.0   
4                                0.000000                            0.0   
..                                    ...                            ...   
144                             13.953488                           35.0   
145                              0.000000                            7.5   
146                              2.325581                            2.5   
147                              0.000000                            5.0   
148                              0.000000                            0.0   

     Percent_Total_Missing  
0                11.382114  
1                 0.813008  
2                 5.691057  
3                 9.756098  
4                 1.626016  
..                     ...  
144              19.512195  
145               3.252033  
146               1.626016  
147               1.626016  
148               0.000000  

[149 rows x 11 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 4. Data cleaning with QC_RSD (QC relative standard deviation) and the total percentage missing values

</div>

In [8]:
# Remove any features with >=20% QC_RSD and >=10% missing values
PeakTableClean = PeakTablePercentMissing.query('QC_RSD < 20 & Percent_Total_Missing < 10')
peaklist = PeakTableClean['Name'] 
X = DataTable[peaklist]
DataTableClean = pd.merge(DataTable.T[~DataTable.T.index.isin(PeakTable['Name'])].T.reset_index(drop=True), pd.DataFrame(X, columns=peaklist).reset_index(drop=True), left_index=True, right_index=True)

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Data Table post data cleanup

Check the cleaned Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(DataTableClean)</span><br>
</div>

In [9]:
display(DataTableClean)

Idx    SampleID SampleType                   Class     M4     M5     M7  \
0      2    sample_2     Sample          Gastric Cancer    NaN  694.5   37.9   
1      3    sample_3     Sample  Benign Gastric Disease   46.8  483.4  110.1   
2      4    sample_4     Sample                 Healthy   14.0   88.6  170.3   
3      5    sample_5     Sample          Gastric Cancer    8.7  243.2  349.4   
4      6    sample_6     Sample  Benign Gastric Disease   18.7  200.1   37.3   
..   ...         ...        ...                     ...    ...    ...    ...   
118  134  sample_134     Sample  Benign Gastric Disease  128.4  324.6   38.9   
119  135  sample_135     Sample                 Healthy   38.4   58.9   42.3   
120  137  sample_137     Sample          Gastric Cancer   91.9  732.1  492.6   
121  138  sample_138     Sample  Benign Gastric Disease   18.7   40.8   46.5   
122  139  sample_139     Sample                 Healthy   20.9   17.4   12.9   

        M8     M11    M14  ...   M126    M129   M130    M134    M137    M138  \
0    125.7   490.6    NaN  ...   61.7   961.8  100.1  6674.1   938.9  6084.5   
1     85.1  2441.2   29.3  ...   76.2   673.4   42.8   787.2  1163.2   246.0   
2     23.9   140.7   62.9  ...   31.1   651.7   31.3   392.4   370.6   109.3   
3     61.1    48.7   77.8  ...  122.4  2048.9    NaN   973.9   984.0  1037.4   
4    243.7   103.7   52.3  ...   38.9   850.6   75.9     NaN  2916.5  1732.1   
..     ...     ...    ...  ...    ...     ...    ...     ...     ...     ...   
118  162.0   112.7  133.9  ...   37.1  2058.3   51.7  1243.5   896.6  2179.1   
119   62.3    23.2   39.0  ...   14.3   818.0   27.9   324.1   288.4    68.5   
120   95.1   194.7  166.7  ...  609.4  3521.4   96.8  2731.0   683.1  3241.9   
121   25.9    50.7   26.3  ...   12.9   435.4    4.6   207.1    68.0   241.5   
122   12.0    21.7   12.9  ...    4.8   890.7   11.6   338.8   283.6   291.8   

      M142   M144    M148   M149  
0     16.1   29.5   390.7  199.0  
1     32.9   33.2    31.6  195.2  
2     60.5   17.0    91.4   91.6  
3     45.6   35.2   322.3  254.3  
4     10.4   32.6   129.7  207.2  
..     ...    ...     ...    ...  
118    7.1   27.4  1868.7  325.8  
119    1.7   32.3     NaN  145.5  
120  182.3  123.9   317.3  401.7  
121   14.5   27.9    47.8   46.5  
122    1.3   24.6   124.1   28.9  

[123 rows x 56 columns]

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table post data cleanup

Check the cleaned Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTableClean)</span><br>
</div>

In [10]:
display(PeakTableClean)

Idx  Name                        Label     Score      VIP1    Pvalue  \
3      3    M4         1-Methylnicotinamide  0.781881  2.360494  0.877783   
4      4    M5               2-Aminoadipate  0.412730  2.969282  0.312104   
6      6    M7              2-Furoylglycine  0.621606  2.910225  0.422395   
7      7    M8         2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   
10    10   M11           3-Aminoisobutyrate  0.052613  1.232261  0.911207   
13    13   M14         3-Hydroxyisobutyrate  0.742954  1.030032  0.040951   
14    14   M15         3-Hydroxyisovalerate  0.863174  1.137913  0.021928   
24    24   M25          6-Hydroxynicotinate  0.257923  0.617403  0.347845   
25    25   M26                          ATP  0.329348  0.592646  0.104580   
30    30   M31                      Adipate  0.830317  2.135924  0.635459   
31    31   M32                      Alanine  0.367747  1.012951  0.424025   
32    32   M33                     Anserine  0.040386  2.489143  0.981209   
35    35   M36                   Asparagine  0.833873  2.288612  0.067982   
36    36   M37                      Azelate  0.867025  2.509055  0.109079   
44    44   M45                      Citrate  0.450986  2.609300  0.756118   
47    47   M48                   Creatinine  0.712154  0.539626  0.200390   
49    49   M50                      Ethanol  0.480404  1.008292  0.024705   
50    50   M51                 Ethanolamine  0.307137  1.157240  0.568156   
64    64   M65                Glycylproline  0.010838  2.550983  0.464338   
65    65   M66                    Hippurate  0.334449  0.613396  0.165290   
67    67   M68                    Histidine  0.346816  2.648516  0.854999   
70    70   M71                    Ibuprofen  0.095636  0.201036  0.301732   
72    72   M73             Indole-3-lactate  0.722734  0.677763  0.733736   
73    73   M74                   Isoleucine  0.415821  1.197205  0.454408   
74    74   M75                      Lactate  0.970615  2.464665  0.959931   
87    87   M88            N-Acetylglutamine  0.851489  0.719889  0.460459   
88    88   M89  N-AcetylglutamineDerivative  0.160368  1.021713  0.497826   
89    89   M90            N-Acetylornithine  0.613409  0.859984  0.617857   
90    90   M91            N-Acetylserotonin  0.920108  1.055689  0.478898   
92    92   M93            N-Methylhydantoin  0.406982  1.071867  0.010548   
100  100  M101                 Pantothenate  0.217481  1.561031  0.355870   
103  103  M104                      Proline  0.602649  0.508679  0.814823   
104  104  M105             Propylene glycol  0.540107  1.765025  0.524586   
105  105  M106                   Pyridoxine  0.844385  2.626348  0.444797   
106  106  M107                Pyroglutamate  0.984642  1.529982  0.414689   
109  109  M110                    Serotonin  0.299790  0.509262  0.852831   
114  114  M115                 Trigonelline  0.623659  0.187733  0.492383   
115  115  M116               Trimethylamine  0.962713  0.356968  0.950067   
117  117  M118                      Tropate  0.015984  1.961317  0.285254   
118  118  M119                   Tryptophan  0.389974  2.900513  0.509550   
119  119  M120                     Tyrosine  0.571739  1.580797  0.383536   
121  121  M122                       Valine  0.151142  1.412871  0.117712   
125  125  M126              trans-Aconitate  0.906070  1.968585  0.026600   
128  128  M129                          u11  0.927506  0.097612  0.682062   
129  129  M130                        u1125  0.560287  2.565725  0.937773   
133  133  M134                         u144  0.562207  2.953544  0.957933   
136  136  M137                         u217  0.933080  0.891412  0.359028   
137  137  M138                         u233  0.648819  1.691548  0.644515   
141  141  M142                          u43  0.481533  2.794356  0.403690   
143  143  M144                          u87  0.090389  2.602484  0.910527   
147  147  M148            π-Methylhistidine  0.116367  2.554018  0.839549   
1

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 5. Statistical analysis

Statistical analysis is important to identify any features or samples which may be outliers. 
It is also important to identify whether the data is normally distributed prior to any further analysis such as correlation analysis. Whether the data is normally distrubuted or not can determine the most suitable correlation function to use. For example the parametric method Pearson's correlation should be used for normally distributed data, whereas the non-parametric method Spearman's correlation is suitable for non-normally distributed data.

Statistical analysis can also provide additional univariate information for futher down-stream visualisations, such as one-way Anova p-values and PCA loadings for each feature to displayed in each of the nodes of the hierarchical edge bundle.
</div>

In [11]:
stats = multivis.utils.statistics(PeakTableClean, DataTableClean)

stats.help()

Generate a table of parametric or non-parametric statistics and merges them with the Peak Table (node table).
        Initial_Parameters
            ----------
            peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
            datatable : Pandas dataframe matrix containing values for statistical analysis

        Methods
            -------
            set_params : Set parameters -
                parametric: Perform parametric statistical analysis, assuming the data is normally distributed (default: True)
                log_data: Perform a log ('natural', base 2 or base 10) on all data prior to statistical analysis (default: (False, 2))
                scale_data: Scale the data to unit variance (default: False)
                impute_data: Impute any missing values using KNN impute with a set number of nearest neighbours (default: (False, 3))
                group_column_name: The group column name used in the datatable (default: None)
      

In [12]:
params = dict({'parametric': False
              , 'log_data': (True, 2)
              , 'scale_data': True
              , 'impute_data': (True, 3)
              , 'group_column_name': 'Class'
              , 'control_group_name': 'Healthy'
              , 'group_alpha_CI': 0.05
              , 'fold_change_alpha_CI': 0.05
              , 'pca_alpha_CI': 0.05
              , 'total_missing': False
              , 'group_missing': False
              , 'pca_loadings': True
              , 'normality_test': True
              , 'group_normality_test': False
              , 'group_mean_CI': False
              , 'group_median_CI': True
              , 'mean_fold_change': False
              , 'median_fold_change': True
              , 'kruskal_wallis_test': False
              , 'levene_twoGroup': False
              , 'levene_allGroup': False
              , 'oneway_Anova_test': False
              , 'ttest_oneGroup': False
              , 'ttest_twoGroup': False
              , 'mann_whitney_u_test': False})

stats.set_params(**params)

PeakTableStats = stats.calculate()

/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used extremal samples; results are probably unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_

/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alph

/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alph

/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=g

/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alpha=grpAlphaCI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:264: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  CIs = bootstrap.ci(data=groupList, statfunction=medianFold, n_samples=500, alpha=fold_change_alpha_CI)
/Users/brettchapman/multivis/multivis/utils/statistics.py:832: InstabilityWarning: Some values used top 10 low/high samples; results may be unstable.
  grpMedian_CI = bootstrap.ci(data=group, statfunction=np.nanmedian, n_samples=500, alph

/Users/brettchapman/multivis/multivis/utils/statistics.py:505: InstabilityWarning: Some values used extremal samples; results are probably unstable.
  PC2_CIs = bootstrap.ci(data=d_filled, statfunction=bootpc2, n_samples=500, alpha=pca_alpha_CI)


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Peak Table with statistical information: Shapiro-Wilk's pvalue, median concentration, median fold change and PCA loadings

Check the imported Peak Table with with statistical information by simply calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(PeakTableStats)</span><br>
</div>

In [13]:
display(PeakTableStats)

Idx  Name                        Label     Score      VIP1    Pvalue  \
0     0    M4         1-Methylnicotinamide  0.781881  2.360494  0.877783   
1     1    M5               2-Aminoadipate  0.412730  2.969282  0.312104   
2     2    M7              2-Furoylglycine  0.621606  2.910225  0.422395   
3     3    M8         2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   
4     4   M11           3-Aminoisobutyrate  0.052613  1.232261  0.911207   
5     5   M14         3-Hydroxyisobutyrate  0.742954  1.030032  0.040951   
6     6   M15         3-Hydroxyisovalerate  0.863174  1.137913  0.021928   
7     7   M25          6-Hydroxynicotinate  0.257923  0.617403  0.347845   
8     8   M26                          ATP  0.329348  0.592646  0.104580   
9     9   M31                      Adipate  0.830317  2.135924  0.635459   
10   10   M32                      Alanine  0.367747  1.012951  0.424025   
11   11   M33                     Anserine  0.040386  2.489143  0.981209   
12   12   M36                   Asparagine  0.833873  2.288612  0.067982   
13   13   M37                      Azelate  0.867025  2.509055  0.109079   
14   14   M45                      Citrate  0.450986  2.609300  0.756118   
15   15   M48                   Creatinine  0.712154  0.539626  0.200390   
16   16   M50                      Ethanol  0.480404  1.008292  0.024705   
17   17   M51                 Ethanolamine  0.307137  1.157240  0.568156   
18   18   M65                Glycylproline  0.010838  2.550983  0.464338   
19   19   M66                    Hippurate  0.334449  0.613396  0.165290   
20   20   M68                    Histidine  0.346816  2.648516  0.854999   
21   21   M71                    Ibuprofen  0.095636  0.201036  0.301732   
22   22   M73             Indole-3-lactate  0.722734  0.677763  0.733736   
23   23   M74                   Isoleucine  0.415821  1.197205  0.454408   
24   24   M75                      Lactate  0.970615  2.464665  0.959931   
25   25   M88            N-Acetylglutamine  0.851489  0.719889  0.460459   
26   26   M89  N-AcetylglutamineDerivative  0.160368  1.021713  0.497826   
27   27   M90            N-Acetylornithine  0.613409  0.859984  0.617857   
28   28   M91            N-Acetylserotonin  0.920108  1.055689  0.478898   
29   29   M93            N-Methylhydantoin  0.406982  1.071867  0.010548   
30   30  M101                 Pantothenate  0.217481  1.561031  0.355870   
31   31  M104                      Proline  0.602649  0.508679  0.814823   
32   32  M105             Propylene glycol  0.540107  1.765025  0.524586   
33   33  M106                   Pyridoxine  0.844385  2.626348  0.444797   
34   34  M107                Pyroglutamate  0.984642  1.529982  0.414689   
35   35  M110                    Serotonin  0.299790  0.509262  0.852831   
36   36  M115                 Trigonelline  0.623659  0.187733  0.492383   
37   37  M116               Trimethylamine  0.962713  0.356968  0.950067   
38   38  M118                      Tropate  0.015984  1.961317  0.285254   
39   39  M119                   Tryptophan  0.389974  2.900513  0.509550   
40   40  M120                     Tyrosine  0.571739  1.580797  0.383536   
41   41  M122                       Valine  0.151142  1.412871  0.117712   
42   42  M126              trans-Aconitate  0.906070  1.968585  0.026600   
43   43  M129                          u11  0.927506  0.097612  0.682062   
44   44  M130                        u1125  0.560287  2.565725  0.937773   
45   45  M134                         u144  0.562207  2.953544  0.957933   
46   46  M137                         u217  0.933080  0.891412  0.359028   
47   47  M138                         u233  0.648819  1.691548  0.644515   
48   48  M142                          u43  0.481533  2.794356  0.403690   
49   49  M144                          u87  0.090389  2.602484  0.910527   
50   50  M148            π-Methylhistidine  0.116367  2.554018  0.839549   
51   51  M149            τ-Methylhistidine  0.870186

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Determine if the data is normally distributed

If the majority of the data shows a Shapiro-Wilks pvalue > 0.05 the data is normally distributed and parametric statistical analysis is recommended, however if the pvalue is < 0.05 then the data is non-normally distributed and non-parametric analysis methods are recommended.
</div>

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Features normally distributed

In [14]:
display(len(PeakTableStats.query('Shapiro_pvalue > 0.05')))

21

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Features non-normally distributed

In [15]:
display(len(PeakTableStats.query('Shapiro_pvalue < 0.05')))

31

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Normality conclusion

The data is non-normally distributed and therefore non-parametric analysis (e.g. Kruskal-Wallis, Mann-Whitney U test, Spearman's correlation etc) is recommended.

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 6. Impute remaining missing values prior to Spearman's correlation analysis
    
Transformation can be done with a log which is commonly used for biological data types, however there are other types such as square, square root, cube root or reciprocal transformation, which may be suitable for other types of data. Scaling is highly dependent on the values found in the dataset. If the dataset contains a number of values which are very large compared to the majority of other values, then this may skew the data in favour of those values and consequently bias the results. Scaling methods such as unit variance scaling scale all the values in the dataset so that the values are all comparable. Pearson correlation, a parametric method, can later be performed where a log is necessary to normally distribute the data. However, if a non-parametric method is used, such as Spearman or Kendall's Tau, then a log is not necessary. Additionally, where a correlation is used no scaling is necessary after log transformation, as correlation analysis does not require scaling, as covariance is measured between individual values in the bivariate correlation analysis, unlike with PCA in tutorial 2.1, which performs multivariate analysis, measuring the variance across all values. However, if another similarity metric were to be used in place of correlation, such as Euclidean distance, then scaling may be a necessary step to take.
    
In this example, as Spearman's correlation is later being performed, no log or scaling is necessary, only the imputation of any remaining missing values is needed.

</div>

In [16]:
peaklist = PeakTableStats['Name']                  # Set peaklist to the metabolite names in the DataTableClean
X = DataTableClean[peaklist]                       # Extract X matrix from DataTableClean using peaklist
#Xlog = np.log10(X)                                # Log transform (base-10)
#Xscale = multivis.utils.scaler(Xlog)              # Scale to unit variance
X = multivis.utils.imputeData(X, k=3)              # Impute remaining missing values using KNN impute with k=3

DataTableClean = pd.merge(DataTableClean.T[~DataTableClean.T.index.isin(PeakTable['Name'])].T.reset_index(drop=True), pd.DataFrame(X, columns=peaklist).reset_index(drop=True), left_index=True, right_index=True)

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 7. Slice up the data into blocks placed into a dictionary indexed by group/class name

Slice the data by group/class name for later identification of multi-block associations and place in a dictionary indexed by group/class name.
</div>

In [17]:
GroupByBlockPeaks, GroupByBlockData = multivis.utils.groups2blocks(PeakTableStats, DataTableClean, 'Class')

In [18]:
keys = GroupByBlockPeaks.keys()
for key in keys:
    display(GroupByBlockPeaks[key])

Idx Name                        Label     Score      VIP1    Pvalue  \
0     0   A1         1-Methylnicotinamide  0.781881  2.360494  0.877783   
1     1   A2               2-Aminoadipate  0.412730  2.969282  0.312104   
2     2   A3              2-Furoylglycine  0.621606  2.910225  0.422395   
3     3   A4         2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   
4     4   A5           3-Aminoisobutyrate  0.052613  1.232261  0.911207   
5     5   A6         3-Hydroxyisobutyrate  0.742954  1.030032  0.040951   
6     6   A7         3-Hydroxyisovalerate  0.863174  1.137913  0.021928   
7     7   A8          6-Hydroxynicotinate  0.257923  0.617403  0.347845   
8     8   A9                          ATP  0.329348  0.592646  0.104580   
9     9  A10                      Adipate  0.830317  2.135924  0.635459   
10   10  A11                      Alanine  0.367747  1.012951  0.424025   
11   11  A12                     Anserine  0.040386  2.489143  0.981209   
12   12  A13                   Asparagine  0.833873  2.288612  0.067982   
13   13  A14                      Azelate  0.867025  2.509055  0.109079   
14   14  A15                      Citrate  0.450986  2.609300  0.756118   
15   15  A16                   Creatinine  0.712154  0.539626  0.200390   
16   16  A17                      Ethanol  0.480404  1.008292  0.024705   
17   17  A18                 Ethanolamine  0.307137  1.157240  0.568156   
18   18  A19                Glycylproline  0.010838  2.550983  0.464338   
19   19  A20                    Hippurate  0.334449  0.613396  0.165290   
20   20  A21                    Histidine  0.346816  2.648516  0.854999   
21   21  A22                    Ibuprofen  0.095636  0.201036  0.301732   
22   22  A23             Indole-3-lactate  0.722734  0.677763  0.733736   
23   23  A24                   Isoleucine  0.415821  1.197205  0.454408   
24   24  A25                      Lactate  0.970615  2.464665  0.959931   
25   25  A26            N-Acetylglutamine  0.851489  0.719889  0.460459   
26   26  A27  N-AcetylglutamineDerivative  0.160368  1.021713  0.497826   
27   27  A28            N-Acetylornithine  0.613409  0.859984  0.617857   
28   28  A29            N-Acetylserotonin  0.920108  1.055689  0.478898   
29   29  A30            N-Methylhydantoin  0.406982  1.071867  0.010548   
30   30  A31                 Pantothenate  0.217481  1.561031  0.355870   
31   31  A32                      Proline  0.602649  0.508679  0.814823   
32   32  A33             Propylene glycol  0.540107  1.765025  0.524586   
33   33  A34                   Pyridoxine  0.844385  2.626348  0.444797   
34   34  A35                Pyroglutamate  0.984642  1.529982  0.414689   
35   35  A36                    Serotonin  0.299790  0.509262  0.852831   
36   36  A37                 Trigonelline  0.623659  0.187733  0.492383   
37   37  A38               Trimethylamine  0.962713  0.356968  0.950067   
38   38  A39                      Tropate  0.015984  1.961317  0.285254   
39   39  A40                   Tryptophan  0.389974  2.900513  0.509550   
40   40  A41                     Tyrosine  0.571739  1.580797  0.383536   
41   41  A42                       Valine  0.151142  1.412871  0.117712   
42   42  A43              trans-Aconitate  0.906070  1.968585  0.026600   
43   43  A44                          u11  0.927506  0.097612  0.682062   
44   44  A45                        u1125  0.560287  2.565725  0.937773   
45   45  A46                         u144  0.562207  2.953544  0.957933   
46   46  A47                         u217  0.933080  0.891412  0.359028   
47   47  A48                         u233  0.648819  1.691548  0.644515   
48   48  A49                          u43  0.481533  2.794356  0.403690   
49   49  A50                          u87  0.090389  2.602484  0.910527   
50   50  A51            π-Methylhistidine  0.116367  2.554018  0.839549   
51   51  A52            τ-Methylhistidine  0.870186  0.739459  0.445225   

       QC_RSD  Percent_Group

Idx Name                        Label     Score      VIP1    Pvalue  \
0     0   B1         1-Methylnicotinamide  0.781881  2.360494  0.877783   
1     1   B2               2-Aminoadipate  0.412730  2.969282  0.312104   
2     2   B3              2-Furoylglycine  0.621606  2.910225  0.422395   
3     3   B4         2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   
4     4   B5           3-Aminoisobutyrate  0.052613  1.232261  0.911207   
5     5   B6         3-Hydroxyisobutyrate  0.742954  1.030032  0.040951   
6     6   B7         3-Hydroxyisovalerate  0.863174  1.137913  0.021928   
7     7   B8          6-Hydroxynicotinate  0.257923  0.617403  0.347845   
8     8   B9                          ATP  0.329348  0.592646  0.104580   
9     9  B10                      Adipate  0.830317  2.135924  0.635459   
10   10  B11                      Alanine  0.367747  1.012951  0.424025   
11   11  B12                     Anserine  0.040386  2.489143  0.981209   
12   12  B13                   Asparagine  0.833873  2.288612  0.067982   
13   13  B14                      Azelate  0.867025  2.509055  0.109079   
14   14  B15                      Citrate  0.450986  2.609300  0.756118   
15   15  B16                   Creatinine  0.712154  0.539626  0.200390   
16   16  B17                      Ethanol  0.480404  1.008292  0.024705   
17   17  B18                 Ethanolamine  0.307137  1.157240  0.568156   
18   18  B19                Glycylproline  0.010838  2.550983  0.464338   
19   19  B20                    Hippurate  0.334449  0.613396  0.165290   
20   20  B21                    Histidine  0.346816  2.648516  0.854999   
21   21  B22                    Ibuprofen  0.095636  0.201036  0.301732   
22   22  B23             Indole-3-lactate  0.722734  0.677763  0.733736   
23   23  B24                   Isoleucine  0.415821  1.197205  0.454408   
24   24  B25                      Lactate  0.970615  2.464665  0.959931   
25   25  B26            N-Acetylglutamine  0.851489  0.719889  0.460459   
26   26  B27  N-AcetylglutamineDerivative  0.160368  1.021713  0.497826   
27   27  B28            N-Acetylornithine  0.613409  0.859984  0.617857   
28   28  B29            N-Acetylserotonin  0.920108  1.055689  0.478898   
29   29  B30            N-Methylhydantoin  0.406982  1.071867  0.010548   
30   30  B31                 Pantothenate  0.217481  1.561031  0.355870   
31   31  B32                      Proline  0.602649  0.508679  0.814823   
32   32  B33             Propylene glycol  0.540107  1.765025  0.524586   
33   33  B34                   Pyridoxine  0.844385  2.626348  0.444797   
34   34  B35                Pyroglutamate  0.984642  1.529982  0.414689   
35   35  B36                    Serotonin  0.299790  0.509262  0.852831   
36   36  B37                 Trigonelline  0.623659  0.187733  0.492383   
37   37  B38               Trimethylamine  0.962713  0.356968  0.950067   
38   38  B39                      Tropate  0.015984  1.961317  0.285254   
39   39  B40                   Tryptophan  0.389974  2.900513  0.509550   
40   40  B41                     Tyrosine  0.571739  1.580797  0.383536   
41   41  B42                       Valine  0.151142  1.412871  0.117712   
42   42  B43              trans-Aconitate  0.906070  1.968585  0.026600   
43   43  B44                          u11  0.927506  0.097612  0.682062   
44   44  B45                        u1125  0.560287  2.565725  0.937773   
45   45  B46                         u144  0.562207  2.953544  0.957933   
46   46  B47                         u217  0.933080  0.891412  0.359028   
47   47  B48                         u233  0.648819  1.691548  0.644515   
48   48  B49                          u43  0.481533  2.794356  0.403690   
49   49  B50                          u87  0.090389  2.602484  0.910527   
50   50  B51            π-Methylhistidine  0.116367  2.554018  0.839549   
51   51  B52            τ-Methylhistidine  0.870186  0.739459  0.445225   

       QC_RSD  Percent_Group

Idx Name                        Label     Score      VIP1    Pvalue  \
0     0   C1         1-Methylnicotinamide  0.781881  2.360494  0.877783   
1     1   C2               2-Aminoadipate  0.412730  2.969282  0.312104   
2     2   C3              2-Furoylglycine  0.621606  2.910225  0.422395   
3     3   C4         2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   
4     4   C5           3-Aminoisobutyrate  0.052613  1.232261  0.911207   
5     5   C6         3-Hydroxyisobutyrate  0.742954  1.030032  0.040951   
6     6   C7         3-Hydroxyisovalerate  0.863174  1.137913  0.021928   
7     7   C8          6-Hydroxynicotinate  0.257923  0.617403  0.347845   
8     8   C9                          ATP  0.329348  0.592646  0.104580   
9     9  C10                      Adipate  0.830317  2.135924  0.635459   
10   10  C11                      Alanine  0.367747  1.012951  0.424025   
11   11  C12                     Anserine  0.040386  2.489143  0.981209   
12   12  C13                   Asparagine  0.833873  2.288612  0.067982   
13   13  C14                      Azelate  0.867025  2.509055  0.109079   
14   14  C15                      Citrate  0.450986  2.609300  0.756118   
15   15  C16                   Creatinine  0.712154  0.539626  0.200390   
16   16  C17                      Ethanol  0.480404  1.008292  0.024705   
17   17  C18                 Ethanolamine  0.307137  1.157240  0.568156   
18   18  C19                Glycylproline  0.010838  2.550983  0.464338   
19   19  C20                    Hippurate  0.334449  0.613396  0.165290   
20   20  C21                    Histidine  0.346816  2.648516  0.854999   
21   21  C22                    Ibuprofen  0.095636  0.201036  0.301732   
22   22  C23             Indole-3-lactate  0.722734  0.677763  0.733736   
23   23  C24                   Isoleucine  0.415821  1.197205  0.454408   
24   24  C25                      Lactate  0.970615  2.464665  0.959931   
25   25  C26            N-Acetylglutamine  0.851489  0.719889  0.460459   
26   26  C27  N-AcetylglutamineDerivative  0.160368  1.021713  0.497826   
27   27  C28            N-Acetylornithine  0.613409  0.859984  0.617857   
28   28  C29            N-Acetylserotonin  0.920108  1.055689  0.478898   
29   29  C30            N-Methylhydantoin  0.406982  1.071867  0.010548   
30   30  C31                 Pantothenate  0.217481  1.561031  0.355870   
31   31  C32                      Proline  0.602649  0.508679  0.814823   
32   32  C33             Propylene glycol  0.540107  1.765025  0.524586   
33   33  C34                   Pyridoxine  0.844385  2.626348  0.444797   
34   34  C35                Pyroglutamate  0.984642  1.529982  0.414689   
35   35  C36                    Serotonin  0.299790  0.509262  0.852831   
36   36  C37                 Trigonelline  0.623659  0.187733  0.492383   
37   37  C38               Trimethylamine  0.962713  0.356968  0.950067   
38   38  C39                      Tropate  0.015984  1.961317  0.285254   
39   39  C40                   Tryptophan  0.389974  2.900513  0.509550   
40   40  C41                     Tyrosine  0.571739  1.580797  0.383536   
41   41  C42                       Valine  0.151142  1.412871  0.117712   
42   42  C43              trans-Aconitate  0.906070  1.968585  0.026600   
43   43  C44                          u11  0.927506  0.097612  0.682062   
44   44  C45                        u1125  0.560287  2.565725  0.937773   
45   45  C46                         u144  0.562207  2.953544  0.957933   
46   46  C47                         u217  0.933080  0.891412  0.359028   
47   47  C48                         u233  0.648819  1.691548  0.644515   
48   48  C49                          u43  0.481533  2.794356  0.403690   
49   49  C50                          u87  0.090389  2.602484  0.910527   
50   50  C51            π-Methylhistidine  0.116367  2.554018  0.839549   
51   51  C52            τ-Methylhistidine  0.870186  0.739459  0.445225   

       QC_RSD  Percent_Group

In [19]:
keys = GroupByBlockData.keys()
for key in keys:
    display(GroupByBlockData[key])

Idx    SampleID SampleType          A1      A2     A3     A4      A5  \
0     2    sample_2     Sample  131.166667   694.5   37.9  125.7   490.6   
1     5    sample_5     Sample    8.700000   243.2  349.4   61.1    48.7   
2     7    sample_7     Sample   38.866667   362.7   59.6   51.3    58.1   
3     9    sample_9     Sample    8.400000   270.2  213.8   65.6    92.9   
4    13   sample_13     Sample   40.900000   106.9    9.5   62.7    58.2   
5    16   sample_16     Sample   10.400000   193.1   72.9   30.7    21.5   
6    20   sample_20     Sample   23.700000    13.4   12.1   47.2   446.6   
7    23   sample_23     Sample    8.000000    38.3   12.2   18.8   153.6   
8    26   sample_26     Sample    6.000000   305.5   53.3   69.2   113.2   
9    30   sample_30     Sample   38.800000   221.0  381.5   43.6    73.6   
10   33   sample_33     Sample   27.600000   160.4   37.5   85.9    55.5   
11   36   sample_36     Sample   10.300000    32.4   20.5   25.3     0.7   
12   40   sample_40     Sample   25.333333   493.5  107.8  204.4   163.8   
13   43   sample_43     Sample   13.800000   180.1   63.8   68.1   562.4   
14   47   sample_47     Sample   55.900000   354.6   63.0  103.2   111.3   
15   50   sample_50     Sample    1.600000   110.2   62.9   65.0    66.8   
16   53   sample_53     Sample   11.200000    25.5  144.2    9.3    20.4   
17   57   sample_57     Sample   50.500000   425.8   78.9  111.8   133.9   
18   58   sample_58     Sample   16.900000   520.0  407.4  137.8   353.2   
19   61   sample_61     Sample   11.000000     2.7  220.9   26.9    51.0   
20   65   sample_65     Sample   15.700000   107.2  116.1   21.0    40.2   
21   68   sample_68     Sample   12.633333    47.4  243.6   17.8    23.8   
22   71   sample_71     Sample   14.133333    23.3   94.3   15.3    47.0   
23   75   sample_75     Sample   26.300000   381.8  295.4   40.6   298.8   
24   78   sample_78     Sample   16.600000   140.8   42.4   37.2    87.2   
25   81   sample_81     Sample    2.100000    32.5    9.1   17.3    33.7   
26   85   sample_85     Sample   19.500000    75.1    9.2   33.4    58.0   
27   88   sample_88     Sample   57.400000   402.7   26.7   35.0  1688.2   
28   92   sample_92     Sample   32.566667   321.6  279.8   23.0    36.2   
29   95   sample_95     Sample   27.700000   514.2  465.0   55.1   205.1   
30   98   sample_98     Sample   14.400000     1.3   31.0   78.5   218.0   
31  102  sample_102     Sample   26.000000    53.9   10.2   48.7    57.0   
32  105  sample_105     Sample   39.900000  2503.0  171.3   98.3   142.5   
33  106  sample_106     Sample   36.800000   333.3   44.0   26.0   307.9   
34  110  sample_110     Sample   56.700000   164.0   18.1   16.5   911.7   
35  113  sample_113     Sample   17.300000    52.2   90.0   26.7    36.6   
36  116  sample_116     Sample   35.900000   256.6   78.7   37.5    29.6   
37  120  sample_120     Sample   22.400000   150.0   59.7   77.6   113.0   
38  123  sample_123     Sample    1.800000   160.3   42.3   17.8    74.7   
39  126  sample_126     Sample   32.800000    94.4   16.3   20.9   607.3   
40  130  sample_130     Sample   24.900000    53.8   38.3   51.8   139.1   
41  133  sample_133     Sample   53.800000    44.6   13.4   34.3   134.0   
42  137  sample_137     Sample   91.900000   732.1  492.6   95.1   194.7   

       A6     A7  ...         A43     A44          A45     A46      A47  \
0    69.5  210.7  ...   61.700000   961.8   100.100000  6674.1    938.9   
1    77.8   51.0  ...  122.400000  2048.9    33.866667   973.9    984.0   
2    34.6   58.1  ...   49.400000  2229.2    61.500000   695.5   1669.3   
3    61.9   54.2  ...   51.500000  2658.3    19.200000  1720.0    368.4   
4   123.6   83.1  ...   32.600000  1800.8    49.900000   766.8    443.7   
5    25.5   91.9  ...   25.200000   519.4     8.000000   597.8     21.5   
6    64.9   67.4  ...   25.100000  1200.9    56.800000   718.1    354.5   
7    50.3   18.5  ...  119.300000   593.1    14.066667   30

Idx    SampleID SampleType          B1     B2          B3     B4  \
0     4    sample_4     Sample   14.000000   88.6  170.300000   23.9   
1     8    sample_8     Sample   18.200000   72.5   15.300000   37.1   
2    12   sample_12     Sample   45.000000   62.6   42.400000   68.0   
3    15   sample_15     Sample   70.600000   65.4   26.200000   81.2   
4    18   sample_18     Sample   13.400000   51.2   23.600000   27.9   
5    22   sample_22     Sample   51.400000  228.6   25.800000  207.4   
6    25   sample_25     Sample   91.800000  551.7  173.400000  112.9   
7    29   sample_29     Sample   58.500000  187.8  166.300000   53.6   
8    32   sample_32     Sample   58.300000  227.2   29.700000  118.0   
9    35   sample_35     Sample   70.300000   43.7   28.300000   99.0   
10   39   sample_39     Sample  109.800000  267.5  316.600000   81.6   
11   42   sample_42     Sample    8.700000   25.8   10.500000    9.3   
12   45   sample_45     Sample   72.700000   67.3   16.300000   50.4   
13   49   sample_49     Sample    7.500000   22.3   13.200000   40.0   
14   52   sample_52     Sample   14.133333   31.3    9.400000   19.5   
15   56   sample_56     Sample    9.533333   45.2   25.100000   22.3   
16   60   sample_60     Sample   49.400000  237.7   28.000000  121.1   
17   63   sample_63     Sample   71.400000  452.3   45.700000  161.0   
18   67   sample_67     Sample  109.400000  510.8  276.100000  126.2   
19   70   sample_70     Sample   45.700000   65.0   17.100000   49.3   
20   74   sample_74     Sample  136.300000  236.2   47.200000  266.6   
21   77   sample_77     Sample  141.800000  360.5   38.800000   57.7   
22   80   sample_80     Sample   16.000000   48.2   13.200000   40.7   
23   84   sample_84     Sample   25.400000  125.1   39.300000   41.2   
24   87   sample_87     Sample   79.800000  252.4   18.000000   71.4   
25   90   sample_90     Sample  144.800000  460.0   54.300000  126.1   
26   94   sample_94     Sample   46.200000   92.7    8.600000   47.7   
27   97   sample_97     Sample   41.400000  175.3   10.800000   49.7   
28  101  sample_101     Sample   52.500000  170.1   45.300000  118.4   
29  104  sample_104     Sample   24.800000   35.2   12.700000   17.3   
30  108  sample_108     Sample    7.800000  143.5   12.100000   26.7   
31  112  sample_112     Sample   39.900000   54.0  152.500000  281.1   
32  115  sample_115     Sample   15.500000   72.8   16.600000   51.2   
33  119  sample_119     Sample   49.000000   77.5   19.033333  114.2   
34  122  sample_122     Sample   19.400000   79.2   13.800000   66.2   
35  125  sample_125     Sample   86.400000  153.5   31.400000   89.2   
36  129  sample_129     Sample    7.300000  822.9   52.800000   99.8   
37  132  sample_132     Sample    6.400000   56.7   23.600000   21.5   
38  135  sample_135     Sample   38.400000   58.9   42.300000   62.3   
39  139  sample_139     Sample   20.900000   17.4   12.900000   12.0   

             B5     B6     B7  ...    B43     B44         B45          B46  \
0    140.700000   62.9   38.3  ...   31.1   651.7   31.300000   392.400000   
1     54.100000   30.3   19.2  ...   20.0  1067.7   57.000000   887.700000   
2    100.700000   45.5   60.8  ...   46.3  1303.2   12.900000   763.700000   
3     73.700000   95.8   48.8  ...   50.6  2833.0    2.500000  1203.700000   
4     58.200000   25.4   24.8  ...    9.1   300.1    5.300000   153.700000   
5    482.300000  295.8  157.0  ...  100.1  2590.6  664.900000  1966.400000   
6    105.100000  132.4  212.3  ...   57.8  3208.8   33.200000  1064.800000   
7      5.700000   92.2   70.7  ...   38.2  3209.2   98.600000   874.700000   
8     77.300000   98.6  108.0  ...   34.3  2512.8    1.500000   722.900000   
9    295.600000  105.2  109.7  ...   37.7  3080.2   34.400000   903.400000   
10    59.400000   39.1   55.0  ...   79.1  2227.5   12.800000    64.600000   
11    20.600000   13.3    7.9  ...    7.9   133.3    1.000000    83.600000   
12   174.000000  114.1   44.3  ...   

Idx    SampleID SampleType          C1           C2          C3     C4  \
0     3    sample_3     Sample   46.800000   483.400000  110.100000   85.1   
1     6    sample_6     Sample   18.700000   200.100000   37.300000  243.7   
2    11   sample_11     Sample   47.400000   146.500000   20.600000   64.4   
3    14   sample_14     Sample   26.000000    95.500000   10.400000   26.4   
4    17   sample_17     Sample   51.100000    45.000000   58.000000   70.5   
5    21   sample_21     Sample   28.100000   353.800000  112.300000   73.3   
6    24   sample_24     Sample   31.000000  2014.700000  370.700000   80.1   
7    27   sample_27     Sample   27.800000   253.300000   16.400000   36.8   
8    31   sample_31     Sample  242.500000   170.100000   74.000000   72.2   
9    34   sample_34     Sample   43.700000   211.300000    8.000000   52.3   
10   38   sample_38     Sample   48.800000   119.600000   70.800000   81.7   
11   41   sample_41     Sample   50.100000    81.300000   35.500000   66.1   
12   44   sample_44     Sample    8.200000    30.900000  137.066667   12.9   
13   48   sample_48     Sample   33.400000    67.000000  204.300000   41.5   
14   51   sample_51     Sample   84.500000   368.400000   28.900000  121.9   
15   54   sample_54     Sample   57.100000    77.200000   19.800000   21.8   
16   59   sample_59     Sample   15.200000   229.300000   19.200000   66.5   
17   62   sample_62     Sample   39.600000    64.600000   18.100000   62.9   
18   66   sample_66     Sample   26.166667   365.000000   26.100000  117.8   
19   69   sample_69     Sample   19.900000   215.466667   22.200000   55.8   
20   72   sample_72     Sample    0.600000   138.500000   28.000000   54.6   
21   76   sample_76     Sample   90.500000    68.800000   21.900000   44.1   
22   79   sample_79     Sample  173.900000   366.400000  224.400000  129.8   
23   83   sample_83     Sample   79.900000    67.200000    4.600000   49.6   
24   86   sample_86     Sample   37.700000   543.800000  169.900000   73.4   
25   89   sample_89     Sample   40.400000   254.800000   35.000000   73.9   
26   93   sample_93     Sample   56.800000   207.666667   17.500000   22.0   
27   96   sample_96     Sample   45.400000   218.000000   38.600000   61.4   
28   99   sample_99     Sample   18.800000    35.000000   12.700000   54.5   
29  103  sample_103     Sample   13.500000    65.200000   10.200000   49.9   
30  107  sample_107     Sample  100.500000   215.800000   32.200000   67.4   
31  111  sample_111     Sample  203.600000  2143.100000   83.300000  525.0   
32  114  sample_114     Sample   51.300000   165.600000   51.100000   59.0   
33  117  sample_117     Sample   32.700000   304.700000   30.300000   76.0   
34  121  sample_121     Sample    0.100000   117.800000   50.800000   37.9   
35  124  sample_124     Sample   39.700000   131.900000   22.500000   53.6   
36  128  sample_128     Sample   73.900000    67.000000   16.700000   22.4   
37  131  sample_131     Sample   38.300000   167.700000   25.800000   90.6   
38  134  sample_134     Sample  128.400000   324.600000   38.900000  162.0   
39  138  sample_138     Sample   18.700000    40.800000   46.500000   25.9   

             C5          C6     C7  ...         C43     C44    C45     C46  \
0   2441.200000   29.300000   45.4  ...   76.200000   673.4   42.8   787.2   
1    103.700000   52.300000   76.8  ...   38.900000   850.6   75.9   759.0   
2     38.500000   42.000000   62.2  ...    8.200000  1579.2   11.3  1625.1   
3    215.000000   19.100000   39.3  ...   30.100000  1943.1   80.8   254.8   
4    161.900000  159.200000   69.7  ...   30.600000  2570.5   16.6  1646.2   
5    618.466667   56.700000   94.7  ...   31.200000   911.6  168.4  1285.0   
6     38.700000  123.700000   60.1  ...  109.600000  3316.3  104.8  5438.4   
7     25.800000   29.700000   79.0  ...   10.366667  1349.1    9.7   397.9   
8    413.000000   62.900000   73.5  ...   19.600000  1730.6   11.9   925.6   
9     36.800000   61.1000

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 8. Merge all the data
    
Merge all the data from each group/block and consolidate any statistical results generated from the multivis.utils.statistics package in relation to each block for later identification of multi-block associations and visualisation.
    
The multi-block data can be prepared using multivis.utils.groups2blocks given a single dataset separated by a group/class column and then merged together on a common index. This is suitable for when comparing differences between groups of samples which do not have a common sample ID. If the dimensions of the index are different for each group, any sample outliers can be identified through exploratory analysis (e.g. PCA) and removed until the dimensions of each group are the same, or if there are no outliers simply merge the data on the common index values and any samples outside the common index will be removed automatically. When the sample IDs are common between each block of data, such as when analysing time-series data comparing the same individuals over time or the same samples across different 'Omics platforms through a systems biology approach (e.g. transcriptomics, proteomics, metabolomics), the multi-block data from each of the "omics" blocks can be placed into a dictionary indexed by the source name. However, the multi-omics approach does require prior modeling with methods such as multi-block variable influence on orthogonal projections (MB-VIOP) and OnPLS or methods such as <a href="http://mixomics.org/">mixOmics</a>. In such a case merging on a common sample ID should be used instead of merging on the index. A good example of this can be seen in this study on asthma by <a href="https://pubs.acs.org/doi/10.1021/acs.analchem.8b03205">Reinke, S et. al. (2018)</a>, where the same samples are measured across multiple 'omics platforms and then log-transformed, scaled and modelled using MB-VIOP and OnPLS.
</div>

In [20]:
MultiBlockPeaks, MultiBlockData = multivis.utils.mergeBlocks(GroupByBlockPeaks, GroupByBlockData, 'Index')

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Multi-block Peak Table

Check the Multi-block Peak Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(MultiBlockPeaks)</span><br>
</div>

In [21]:
display(MultiBlockPeaks)

Idx Name                 Label     Score      VIP1    Pvalue     QC_RSD  \
0      0   A1  1-Methylnicotinamide  0.781881  2.360494  0.877783  12.804201   
1      1   A2        2-Aminoadipate  0.412730  2.969282  0.312104   9.372664   
2      2   A3       2-Furoylglycine  0.621606  2.910225  0.422395   5.049156   
3      3   A4  2-Hydroxyisobutyrate  0.339914  0.636861  0.040471   5.132340   
4      4   A5    3-Aminoisobutyrate  0.052613  1.232261  0.911207  15.476165   
..   ...  ...                   ...       ...       ...       ...        ...   
151  151  C48                  u233  0.648819  1.691548  0.644515   4.735334   
152  152  C49                   u43  0.481533  2.794356  0.403690   7.151166   
153  153  C50                   u87  0.090389  2.602484  0.910527   6.635486   
154  154  C51     π-Methylhistidine  0.116367  2.554018  0.839549  16.561921   
155  155  C52     τ-Methylhistidine  0.870186  0.739459  0.445225   8.351801   

     Percent_Group_Benign Gastric Disease_Missing  \
0                                             7.5   
1                                             5.0   
2                                             7.5   
3                                             0.0   
4                                            10.0   
..                                            ...   
151                                           0.0   
152                                           0.0   
153                                           0.0   
154                                           0.0   
155                                           0.0   

     Percent_Group_Gastric Cancer_Missing  Percent_Group_Healthy_Missing  ...  \
0                               16.279070                            5.0  ...   
1                                0.000000                            0.0  ...   
2                                0.000000                            2.5  ...   
3                                0.000000                            0.0  ...   
4                                4.651163                            2.5  ...   
..                                    ...                            ...  ...   
151                              0.000000                            2.5  ...   
152                              2.325581                            0.0  ...   
153                              0.000000                            0.0  ...   
154                              0.000000                            5.0  ...   
155                              0.000000                            0.0  ...   

     PC2_sig                   Block  MedianFoldChange  \
0       True          Gastric Cancer         -0.452387   
1      False          Gastric Cancer         -0.486787   
2       True          Gastric Cancer         -0.864664   
3       True          Gastric Cancer         -1.848138   
4      False          Gastric Cancer         -0.190740   
..       ...                     ...               ...   
151     True  Benign Gastric Disease         -0.407792   
152     True  Benign Gastric Disease         -0.104473   
153    False  Benign Gastric Disease          0.702931   
154     True  Benign Gastric Disease         -0.475721   
155    False  Benign Gastric Disease       -161.441077   

     MedianFoldChange_CI_lower  MedianFoldChange_CI_upper  \
0                    -1.926394                   1.290010   
1                   -11.102831                   3.381420   
2                    -6.755540                  -0.022641   
3                  -609.649862                  -0.208958   
4                   -44.452314                   5.179738   
..                         ...                        ...   
151                  -1.004521                  -0.108503   
152                  -1.882861                   1.437876   
153                   0.388747                   1.104359   
154                  -1.539270                   0.276870   
155               -1007.292158                 -11.482951   

    

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display the Multi-block Data Table

Check the Multi-block Data Table simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(MultiBlockData)</span><br>
</div>

In [22]:
display(MultiBlockData)

Idx SampleType          A1      A2     A3     A4      A5     A6     A7  \
0     2     Sample  131.166667   694.5   37.9  125.7   490.6   69.5  210.7   
1     5     Sample    8.700000   243.2  349.4   61.1    48.7   77.8   51.0   
2     7     Sample   38.866667   362.7   59.6   51.3    58.1   34.6   58.1   
3     9     Sample    8.400000   270.2  213.8   65.6    92.9   61.9   54.2   
4    13     Sample   40.900000   106.9    9.5   62.7    58.2  123.6   83.1   
5    16     Sample   10.400000   193.1   72.9   30.7    21.5   25.5   91.9   
6    20     Sample   23.700000    13.4   12.1   47.2   446.6   64.9   67.4   
7    23     Sample    8.000000    38.3   12.2   18.8   153.6   50.3   18.5   
8    26     Sample    6.000000   305.5   53.3   69.2   113.2   50.9  109.1   
9    30     Sample   38.800000   221.0  381.5   43.6    73.6   81.5   59.3   
10   33     Sample   27.600000   160.4   37.5   85.9    55.5   29.1   45.0   
11   36     Sample   10.300000    32.4   20.5   25.3     0.7   73.9   51.8   
12   40     Sample   25.333333   493.5  107.8  204.4   163.8   56.1   79.8   
13   43     Sample   13.800000   180.1   63.8   68.1   562.4  241.5   19.1   
14   47     Sample   55.900000   354.6   63.0  103.2   111.3  176.7  126.7   
15   50     Sample    1.600000   110.2   62.9   65.0    66.8   67.3   53.8   
16   53     Sample   11.200000    25.5  144.2    9.3    20.4   22.5   21.1   
17   57     Sample   50.500000   425.8   78.9  111.8   133.9   44.8   28.4   
18   58     Sample   16.900000   520.0  407.4  137.8   353.2  196.8  195.6   
19   61     Sample   11.000000     2.7  220.9   26.9    51.0   28.3   21.2   
20   65     Sample   15.700000   107.2  116.1   21.0    40.2   23.8   24.0   
21   68     Sample   12.633333    47.4  243.6   17.8    23.8   37.0   21.5   
22   71     Sample   14.133333    23.3   94.3   15.3    47.0   26.7   22.4   
23   75     Sample   26.300000   381.8  295.4   40.6   298.8   45.3   29.3   
24   78     Sample   16.600000   140.8   42.4   37.2    87.2   19.6   31.6   
25   81     Sample    2.100000    32.5    9.1   17.3    33.7   18.6   13.0   
26   85     Sample   19.500000    75.1    9.2   33.4    58.0   34.6   28.2   
27   88     Sample   57.400000   402.7   26.7   35.0  1688.2   50.6   33.2   
28   92     Sample   32.566667   321.6  279.8   23.0    36.2   35.2   24.0   
29   95     Sample   27.700000   514.2  465.0   55.1   205.1   34.5   62.2   
30   98     Sample   14.400000     1.3   31.0   78.5   218.0   50.4   26.1   
31  102     Sample   26.000000    53.9   10.2   48.7    57.0   43.3   33.8   
32  105     Sample   39.900000  2503.0  171.3   98.3   142.5   83.7  163.6   
33  106     Sample   36.800000   333.3   44.0   26.0   307.9   39.0   48.7   
34  110     Sample   56.700000   164.0   18.1   16.5   911.7    0.5   78.2   
35  113     Sample   17.300000    52.2   90.0   26.7    36.6   21.5   25.8   
36  116     Sample   35.900000   256.6   78.7   37.5    29.6   44.1   43.7   
37  120     Sample   22.400000   150.0   59.7   77.6   113.0   94.3   86.3   
38  123     Sample    1.800000   160.3   42.3   17.8    74.7   17.0   37.4   
39  126     Sample   32.800000    94.4   16.3   20.9   607.3   40.6   39.6   

      A8  ...         C43     C44    C45     C46     C47     C48    C49  \
0   28.3  ...   76.200000   673.4   42.8   787.2  1163.2   246.0   32.9   
1   58.9  ...   38.900000   850.6   75.9   759.0  2916.5  1732.1   10.4   
2   57.7  ...    8.200000  1579.2   11.3  1625.1   242.1   488.8    6.1   
3   39.6  ...   30.100000  1943.1   80.8   254.8   428.7   840.0  282.9   
4   15.9  ...   30.600000  2570.5   16.6  1646.2   251.7   697.2    8.3   
5   16.1  ...   31.200000   911.6  168.4  1285.0  1389.8  3609.6   44.0   
6   17.1  ...  109.600000  3316.3  104.8  5438.4  2218.0   515.7   77.3   
7    0.8  ...   10.366667  1349.1    9.7   397.9   176.3  2981.7   12.8   
8   27.0  ...   19.600000  1730.6   11.9   925.6  1115.2  1919.7   23.5   
9   64.7  ...   49.300000  2129.0   79.0   252.2   908.

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 9. Correlation analysis

Correlation is a form of similarity and measures the strength of the linear relationship between two variables. Pearson's correlation, a form of parametric correlation analysis, is described mathematically by dividing the joint variability or covariance of two variables by the product of their standard deviations (see Eq1). Other forms of correlation measure the monotonic relationships and are non-parametric, such as Spearman’s rank correlation and Kendall Tau's correlation. The following correlation analysis, allows for Pearson, Spearman or Kendall Tau's correlation analysis.

\begin{equation*}
r = \frac{Cov(X,Y)}{SD(X).SD(Y)}
\end{equation*}
<center>Eq1: Pearson’s correlation coefficient</center>

</div>

In [23]:
correlationType = "spearman" #"pearson"; "kendalltau"

X = MultiBlockData[MultiBlockPeaks['Name']]

MultiBlockScores,MultiBlockPvalues = multivis.utils.corrAnalysis(X, correlationType)

100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 156/156 [00:14<00:00, 10.42it/s]


<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 10.  Generate Edges

The similarities/distances are filtered and put in a dataframe of edges, where nodes represent features (metabolites) and edges represent similarity/distance scores (e.g. correlation coefficients), with included node names, labels and other attributes such as correlation coefficient pvalues and other statistical information.

</div>

In [24]:
networkEdges = multivis.Edge(peaktable=MultiBlockPeaks, datatable=MultiBlockScores, pvalues=MultiBlockPvalues)

networkEdges.help()

Builds nodes and edges and is the base class for the Network class.

        Initial_Parameters
        ----------
        peaktable : Pandas dataframe containing peak data. Must contain 'Name' and 'Label'.
        datatable : Pandas dataframe matrix containing scores
        pvalues : Pandas dataframe matrix containing score/similarity pvalues (if available, otherwise set to None)

        Methods
        -------
        set_params : Set parameters
            filter_type: The value type to filter the data on (default: 'pvalue')
            hard_threshold: Value to filter the data on (default: 0.005)
            withinBlocks: Include scores within blocks if building multi-block network (default: False)
            sign: The sign of the score/similarity to filter on ('pos', 'neg' or 'both') (default: 'both')

        help : Print this help text

        build : Builds the nodes and edges.
        getNodes : Returns a Pandas dataframe of all nodes.
        getEdges : Returns a Pandas da

In [25]:
params = dict({'filter_type': 'pvalue'               #The filer type to use for the similarities matrix ('Pvalue' or 'Score')              
                    , 'hard_threshold': 0.05          #The hard threshold to apply to the similarities matrix
                    , 'withinBlocks': True          #Include scores within blocks if building multi-block network
                    , 'sign': "both"})               #The sign of the similarities ('pos', 'neg' or 'both')

networkEdges.set_params(**params)

networkEdges.build()

edges = networkEdges.getEdges()
nodes = networkEdges.getNodes()

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display node data used in the Hierarchical Edge Bundle

Check the node data simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(nodes)</span><br>
</div>

In [26]:
display(nodes)

Idx Name                 Label                Score                VIP1  \
0      0   A1  1-Methylnicotinamide   0.7818809592495104  2.3604940845644222   
1      1   A2        2-Aminoadipate  0.41272975250933697   2.969282495538328   
2      2   A3       2-Furoylglycine   0.6216059593217798  2.9102253505330196   
3      3   A4  2-Hydroxyisobutyrate  0.33991383587817403  0.6368605295964379   
4      4   A5    3-Aminoisobutyrate  0.05261303702988451  1.2322610580806348   
..   ...  ...                   ...                  ...                 ...   
151  151  C48                  u233   0.6488194992335996  1.6915475902572799   
152  152  C49                   u43  0.48153258586917613  2.7943561790255034   
153  153  C50                   u87   0.0903890065459535  2.6024844722124936   
154  154  C51     π-Methylhistidine  0.11636744927699838  2.5540178928384805   
155  155  C52     τ-Methylhistidine   0.8701858668724608  0.7394589178223214   

                  Pvalue            QC_RSD  \
0     0.8777833638996088  12.8042008738419   
1     0.3121038626986239  9.37266355735233   
2     0.4223951783297637  5.04915562556207   
3    0.04047086649009335  5.13234006385694   
4     0.9112069804740864  15.4761649136304   
..                   ...               ...   
151   0.6445147893501302  4.73533373690427   
152  0.40369030143761075  7.15116605857213   
153   0.9105266138515601  6.63548612638398   
154   0.8395491475231609  16.5619208190305   
155  0.44522519478107525   8.3518014361243   

    Percent_Group_Benign Gastric Disease_Missing  \
0                                            7.5   
1                                            5.0   
2                                            7.5   
3                                            0.0   
4                                           10.0   
..                                           ...   
151                                          0.0   
152                                          0.0   
153                                          0.0   
154                                          0.0   
155                                          0.0   

    Percent_Group_Gastric Cancer_Missing Percent_Group_Healthy_Missing  ...  \
0                      16.27906976744186                           5.0  ...   
1                                    0.0                           0.0  ...   
2                                    0.0                           2.5  ...   
3                                    0.0                           0.0  ...   
4                      4.651162790697675                           2.5  ...   
..                                   ...                           ...  ...   
151                                  0.0                           2.5  ...   
152                   2.3255813953488373                           0.0  ...   
153                                  0.0                           0.0  ...   
154                                  0.0                           5.0  ...   
155                                  0.0                           0.0  ...   

    PC2_sig                   Block      MedianFoldChange  \
0      True          Gastric Cancer   -0.4523869055976204   
1     False          Gastric Cancer   -0.4867867778878819   
2      True          Gastric Cancer   -0.8646635029368777   
3      True          Gastric Cancer    -1.848137734675045   
4     False          Gastric Cancer  -0.19074018223238542   
..      ...                     ...                   ...   
151    True  Benign Gastric Disease   -0.4077922966405597   
152    True  Benign Gastric Disease  -0.10447290354913867   
153   False  Benign Gastric Disease    0.7029307289750731   
154    True  Benign Gastric Disease   -0.4757207117363398   
155   False  Benign Gastric Disease    -161.4410768482232   

    MedianFoldChange_CI_lower MedianFoldChange_CI_upper MedianFoldChange_sig  \
0         -1.9263939899895632        1.2900098311142276                False   
1         -11.1028305478

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Display edge data used in the Hierarchical Edge Bundle

Check the edge data simply by calling the function <span style="font-family: monaco; font-size: 14px; background-color:white;">display(edges)</span><br>
</div>

In [27]:
display(edges)

start_index start_name           start_label             start_block  \
0               0         A1  1-Methylnicotinamide          Gastric Cancer   
1               0         A1  1-Methylnicotinamide          Gastric Cancer   
2               0         A1  1-Methylnicotinamide          Gastric Cancer   
3               0         A1  1-Methylnicotinamide          Gastric Cancer   
4               0         A1  1-Methylnicotinamide          Gastric Cancer   
...           ...        ...                   ...                     ...   
3108          150        C47                  u217  Benign Gastric Disease   
3109          150        C47                  u217  Benign Gastric Disease   
3110          151        C48                  u233  Benign Gastric Disease   
3111          151        C48                  u233  Benign Gastric Disease   
3112          152        C49                   u43  Benign Gastric Disease   

      end_index end_name             end_label               end_block  \
0             1       A2        2-Aminoadipate          Gastric Cancer   
1             4       A5    3-Aminoisobutyrate          Gastric Cancer   
2             6       A7  3-Hydroxyisovalerate          Gastric Cancer   
3             7       A8   6-Hydroxynicotinate          Gastric Cancer   
4             8       A9                   ATP          Gastric Cancer   
...         ...      ...                   ...                     ...   
3108        153      C50                   u87  Benign Gastric Disease   
3109        154      C51     π-Methylhistidine  Benign Gastric Disease   
3110        154      C51     π-Methylhistidine  Benign Gastric Disease   
3111        155      C52     τ-Methylhistidine  Benign Gastric Disease   
3112        153      C50                   u87  Benign Gastric Disease   

         score  sign    pvalue  
0     0.513321   1.0  0.000706  
1     0.413696   1.0  0.007966  
2     0.368498   1.0  0.019300  
3     0.437075   1.0  0.004802  
4     0.332083   1.0  0.036305  
...        ...   ...       ...  
3108  0.318205   1.0  0.045388  
3109  0.330957   1.0  0.036981  
3110  0.365666   1.0  0.020322  
3111  0.349906   1.0  0.026876  
3112  0.448090   1.0  0.003737  

[3113 rows x 11 columns]

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 11.  Sort edges

Sort the edges for visualisation preference

</div>

In [28]:
#edges.sort_values(['start_index', 'end_index'], inplace=True, ascending=True)
#edges.sort_values('pvalue', inplace=True, ascending=False)
edges.sort_values('score', inplace=True, ascending=False)

<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 12.  Plot Hierarchical edge bundle

The edges from the network are then passed into D3 JavaScript to generate a Hierarchical edge bundle and embedded in HTML for interactive visualisation. The Hierarchical edge bundle is implemented as a circular hierarchical tree structure, with nodes on the outside and edges passing through the circle following a bundled curve until they connect to other nodes. The edges represent some association value such as correlation coefficients and can be coloured accordingly based on the values, the sign or represented as pvalues and coloured using a continuous colour map. Different meta data such as groups/classes within the data can also be reflected in the plot to illustrate how the different groups/classes are correlated and to what degree.

Note: The visualisation will automatically open in another tab, unless running in Binder (See step 13).

</div>

In [29]:
bundle = multivis.edgeBundle(nodes,edges)

bundle.help()

Produces an interactive hierarchical edge bundle in D3.js, from nodes and edges.

        Parameters
        ----------
        nodes : Pandas dataframe containing nodes generated from Edge.
        edges : Pandas dataframe containing edges generated from Edge.
        
        Methods
        -------
        set_params : Set parameters -
            html_file: Name to save the HTML file as (default: 'hEdgeBundle.html')
            innerRadiusOffset: Sets the inner radius based on the offset value from the canvas width/diameter (default: 120)
            blockSeparation: Value to set the distance between different segmented blocks (default: 1)
            linkFadeOpacity: The link fade opacity when hovering over/clicking nodes (default: 0.05)
            mouseOver: Setting to 'True' swaps from clicking to hovering over nodes to select them (default: True)
            fontSize: The font size in pixels set for each node (default: 10)
            backgroundColor: Set the background colour

In [30]:
params = dict({'html_file': 'hEdgeBundle_multi-block_gastric_cancer.html'      #HTML file name to save to
               , 'innerRadiusOffset': 120           #The offset from the radius to determine the inner radius of the edge bundle
               , 'blockSeparation': 5               #The degree of separation between blocks
               , 'linkFadeOpacity': 0.01            #The opacity of faded links
               , 'mouseOver': True                  #Setting to 'True' swaps from clicking to hovering over nodes to select them 
               , 'fontSize': 10                     #The font size of each node
               , 'backgroundColor': 'white'         #Set the background colour of the plot
               , 'foregroundColor': 'black'         #Set the foreground colour of the plot
               , 'node_data': ['Name', 'Label', 'VIP1', 'Pvalue', 'Group_median', 'Group_median_sig', 'MedianFoldChange', 'MedianFoldChange_sig', 'PC1', 'PC2', 'PC1_sig', 'PC2_sig']
               , 'nodeColorScale': 'linear'         #The scale to use for colouring the nodes
               , 'node_color_column': 'Pvalue'      #If node_color_column contains colour values it overides the use of node_cmap
               , 'node_cmap': 'brg'                 #Set the colour palette to use for colouring the nodes               
               , 'edge_color_value': 'sign'         #Set the values to colour the edges by. Either 'sign', 'score' or 'pvalue'.
               , 'edgeColorScale': 'linear'         #The scale to use for colouring the edges (if edge_color_value is 'pvalue')
               , 'edge_cmap': 'cool'                #Set the colour palette to use for colouring the edges
               , 'addArcs': True                    #Setting to 'True' adds arcs around the edge bundle for each block
               , 'arcRadiusOffset': 20              #Sets the arc radius offset from the inner radius
               , 'extendArcAngle': 2                #Sets the angle value to add to each end of the arcs 
               , 'arc_cmap': 'tab20'})              #Sets the CMAP colour palette to use for colouring the arcs

bundle.set_params(**params)

bundle.build()

HTML writen to hEdgeBundle_multi-block_gastric_cancer.html


<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Build Dashboard

A dashboard with panels for the hierarchical edge bundle, node data and sliders is built, allowing for a more robust interface for exploratory analysis of the data. The dashboard is automatically written to HTML and launched upon creation.

</div>

In [31]:
bundle.buildDashboard()

HTML writen to hEdgeBundle_multi-block_gastric_cancer_dashboard.html


<div style="background-color:rgb(255, 250, 250); padding:10px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

## 13.  Alternative visualisation options

The visualisations will automatically open when run locally in Jupyter notebook. However, when running in Binder the visualisations will not open in a new tab due to security restrictions within Binder. Therefore, the following options are available: IFrames and JavaScript.

Note 1: Due to security restrictions within Jupyter notebook, resizing the window opened from within Jupyter notebook with Javascript or IFrame will result in a 403: Forbidden error, due to the visualisation being reloaded each time. In this case, just change the dimensions set in the cell and rerun the cell.

Note 2: Depending on the browser you're using the save button may be disabled when opening the visualisation in an IFrame or JavaScript pop-up. In Chrome downloads have been disable to prevent malicious behavior. If you have trouble downloading try a different browser or use the visualisation opened automatically in another tab when running Jupyter notebook locally.
</div>

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Plain visualisation
</div>

In [32]:
vis_option = "IFrame"  #"Javascript"
file = params['html_file']

if vis_option.lower() == "javascript":
    display(Javascript('''window.open(\'{}\','edgeBundle','width=1000,height=1000')'''.format(file)))
elif vis_option.lower() == "iframe":
    display(IFrame(src=file, width='100%', height='1000px'))

<div style="background-color:rgb(255, 250, 250); padding:5px;  border: 1px solid lightgrey; padding-left: 1em; padding-right: 1em;">

### Visualisation with dashboard
</div>

In [33]:
vis_option = "javascript" #"IFrame"
file = params['html_file'].split(".")[0]+"_dashboard.html"

if vis_option.lower() == "javascript":
    display(Javascript('''window.open(\'{}\','edgeBundle','width=1500,height=1000')'''.format(file)))
elif vis_option.lower() == "iframe":
    display(IFrame(src=file, width='100%', height='1000px'))

<IPython.core.display.Javascript object>